### DISCUSSION TITLE
How to find the name of an element's bound property

### DISCUSSION THREAD

**Discussion by docdunning** [alpinejs/alpine#4033] [1. Help]
In a custom directive, we have easy access to the DOM element via the `el` parameter.  Given this HTML element:

`<input type="text" x-model.debounce.1000ms="foo">`

I want to be able to get hold of the value "foo", i.e. the name of the bound property.

I know that `el._x_model.get()` returns the **value** of the bound property, but there doesn't appear to be a way to get the **name** of the bound property.  All I've been able to come up with is to loop through the element's attributes, looking for one that starts with `x-model`, but this makes the code more complicated than I'd like.

So - how to get that bound property's name?

**Comment by ekwoka** [comment]
You can't really.

this might be an XY problem though.

what is the thing you ACTUALLY want to do?

**Reply by docdunning** [reply]
> what is the thing you ACTUALLY want to do?

Fair question :-).  I have a form with many fields.  Some of them are significant, in the sense that if they are changed, then something else has to happen.  It's good to use $watch for this, but it gets laborious to add that to each element.

One approach I've tried (as shown in the code below) is to create a custom directive `x-watch`.  This looks for the `x-model` attribute, and sets up a watch on that property.  Hence my question - as you can see, having to loop through the element's attributes is a tad clumsy.

In this example, I'm using the `value` passed into the directive to set a boolean indicator to true when any of the watches is triggered.  So `x-watch:keychange` will set `keychange` to true.

An easy way to avoid the attribute loop is to use (e.g.) `x-watch="message"`.  This seemed unnecessarily repetitive though - the element already knows which property it's bound to.

Btw, I have another approach which doesn't use a directive but sets up the watches from an array of property names in the init() function.  Too many choices!

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.5/dist/cdn.min.js"></script>
</head>
<body>
<div x-data="mydata">
  <div>
    <input x-watch:keychange type="text" x-model.lazy="message">
    <input type="text" x-model.lazy="name">
    <input x-watch:keychange type="text" x-model.debounce.500ms="town">
  </div>
  <p x-text="keychange"></p>
</div>
<script>
document.addEventListener('alpine:init', () => {
  Alpine.directive('watch', (el, { value }, { Alpine }) => {
    for (const attr of el.attributes) {
      if (attr.name.startsWith('x-model')) {
        Alpine.$data(el).$watch(attr.value, () => {
          Alpine.$data(el)[value] = true
        })
      }
    }
  })
  Alpine.data('mydata',() => ({
    message: 'Hello world',
    name: '',
    town: '',
    keychange: false,
  }))
})
</script>
</body>
</html>
```

**Reply by ekwoka** [reply]
This gets a bit internal, but it seems like what you want (in terms of tapping into the model updates) would be handled with tapping into `el._x_model` which is an object of getter/setter used by the x-model.

You can basically take that out and replace it with a new getter/setter object that does your additional logic inside of it.

But you should probably just be using `@change`. Since there will be a change event when the thing changes...since that's all x-model uses anyway.

similarly you could use `x-bind` to just bind x-model and whatever else at the same time like `x-bind="watchmodel('message')"`

**Reply by docdunning** [reply]
Thanks for the reminder about `x-bind`, and you're right that `@change` is probably all I need for this use case.

Here's a page demonstrating two approaches I've tried after your reply.  

I've used a simple `@change` on the third input. It's economical on code, and clear as to what it does, so I think that will be my preferred pattern.

Just for the exercise, I modified my `x-watch` directive to use `Alpine.bind()`.  It works fine, but in this case it incurs more code for no great benefit.  Still, at least I know it can be done.   

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.5/dist/cdn.min.js"></script>
</head>
<body>
<div x-data="mydata">
  <div>
    <input x-watch:keychange type="text" x-model="message">
    <input type="text" x-model.lazy="name">
    <input @change="flag('keychange')" type="text" x-model="town">
  </div>
  <p x-text="keychange"></p>
</div>
<script>
document.addEventListener('alpine:init', () => {
  Alpine.directive('watch', (el, { value }, { Alpine }) => {
    Alpine.bind(el, () => ({
      '@change'() { this[value] = true },
    }))
  })
  Alpine.data('mydata',() => ({
    message: 'Hello world',
    name: '',
    town: '',
    keychange: false,
    flag: function (indicator) { this[indicator] = true },
  }))
})
</script>
</body>
</html>
```

**Reply by ekwoka** [reply]
I meant using bind to add the x-model and do $watch at the same time so you only pass the expression in once.

But yes, I think this specific use case, @change is the most sensible to avoid needless abstraction.

**Reply by docdunning** [reply]
> I meant using bind to add the x-model and do $watch at the same time

Understood - thanks.
